### PR TITLE
Switch to geckodriver/firefox in bokeh.io

### PR DESCRIPTION
**Pull request by mattpap**
This adds support for geckodriver/firefox and chromedriver/chromium, and removes support for obsolete phantomjs. The former pair is now the default, primarily because both geckodriver and firefox are available from conda-forge, so installation process is straightforward. The later pair can be used alternatively. I plan to add support for auto-detection of available drivers and headless browsers, and make bokeh use what's available, so that we don't force users to install new web browsers needlessly. Especially because there should be a viable browser available on modern systems.

### CODE REVIEW

**Comment by bryevdv** [comment]
@mattpap have you compared execution times with the new drivers? Last time I checked against Chromedriver, phantomJS was *much* faster.

**Comment by mattpap** [comment]
Based on my rudimentary experimentation, the average execution times for phantomjs, chromium and firefox are 2.4, 2.6 and 4 seconds respectively, with firefox experiencing greater variation than other browsers. So, chromium could be the default, as it is only marginally slower, but harder to install, and the we could fall back to firefox, which is readily available from conda-forge. I suppose those times could be improved by e.g. configuring an empty profile for both browsers and maybe applying some other optimizations.

**Comment by mattpap** [comment]
Discussing the validity of continuing using phantomjs, one has to also consider that this is an obsolete, unmaintained web browser, which is a growing security risk for our users. I also intend to switch to ES6 resources by default, which will eventually penalize phantomjs and IE in terms of execution times.

**Comment by bryevdv** [comment]
To be clear: I'm not suggesting we have to continue with PhantomJS, but I am saying we need to make sure we:
* understand any trade-offs that are unavoidable so that they can be documented and socialized 
* identify specific areas that might be able to be improved

**Comment by mattpap** [comment]
> I'm also not opposed to leaving PhantomJS as a non-default option for some time, either.

One issue with leaving phantomjs support in `bokeh.io`, is that I intend to switch the default ES target to ES2017, which means that the default bokehjs bundles won't work in phantomjs anymore, and exporting stuff would require conditionally generated HTML depending on the browser. It's possible, though inconvenient. Also just to note, support for phantomjs and IE will still be possible via legacy bundles. I will explain more in a PR when ready.

**Comment by mattpap** [comment]
This took way more effort than I initially thought it would take. I made chromium the default, as it is the faster of the two headless browsers.

**Comment by mattpap** [comment]
Another reason for an overhaul of the resources and related sub-systems:
```
WARNING  bokeh.io.export:export.py:293 There were browser warnings and/or errors that may have affected your export
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_css_1 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_css_2 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_css_3 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/a_class - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_js_1 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_js_2 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/external_js_3 - Failed to load resource: net::ERR_FILE_NOT_FOUND
WARNING  bokeh.io.export:export.py:295 file:///tmp/z_class - Failed to load resource: net::ERR_FILE_NOT_FOUND
```

**Comment by bryevdv** [comment]
Where are those coming from?

**Comment by mattpap** [comment]
Tests in `test_resources` create dummy custom models, which are then picked up by resources when testing exports.

**Comment by mattpap** [comment]
@bryevdv, can you check what's the path to firefox on macos and if it doesn't behave weird at startup (like I found out about chrome and windows in c4ef3f9)? Currently firefox not starting is the only issue left to fix in this PR.

**Comment by bryevdv** [comment]
@mattpap on my machine it is
```
 /Applications/Firefox.app/Contents/MacOS/firefox
```

FYI unrelateds: can you check for some questions you've been tagged in on the Discourse, at some point? Also you should be able to sign on to the Zulip instance now

**Comment by mattpap** [comment]
Looks like I finally figured out how to make this work on all supported platforms. Two days lost, but at least a bunch for problems were resolved:

- can't rely on `chrome --version` on windows
- can't rely on `which chrome` (and any other variation) in general
- `window.outer{Width,Height}` can be zero when `inner{Width,Height}` is positive
- chrome on macos can't preserve colors (e.g. 0xFF0000 (red) becomes 0xFC0006)

The ability to upload artifacts (png and html) was critical for resolving this effectively.

**Comment by mattpap** [comment]
> can you check for some questions you've been tagged in on the Discourse (...)

If I grow another pair of hands, sure. Jokes aside, I will do that at some point, however it would be great if I got notifications at the time I'm being mentioned (like on GitHub). I already configured Discourse to be maximally verbose, and I'm still getting nothing (well, a batch from time to time).

File: sphinx/source/docs/user_guide/export.rst [hunk 1]
```diff
@@ -6,21 +6,18 @@ Exporting Plots
...
-.. code-block:: sh
-
-    npm install -g phantomjs-prebuilt
-
-Note that the minimum compatible version of PhantomJS is 2.1, regardless of
-how it is installed.
+Alternatively one can use ``chromedriver`` and ``chromium`` browser (or one of
+its variants). Typically at least one, compatible browser will be available on
+modern systems, thus installing a web browser may not be necessary. bokeh will
+search for available browsers (and drivers) and use what's available, unless
```
Comment (bryevdv) [inline sphinx/source/docs/user_guide/export.rst]: No comma after "at least one", also "bokeh" shoudl be capitalized in "bokeh will search..."

**Comment by bryevdv** [comment]
> I already configured Discourse to be maximally verbose, and I'm still getting nothing (well, a batch from time to time).

Weird, I seem to get more than digests. I will look in to it. 

Re: this PR, are there any performance comparisons between the new options and phantomjs? We should come up with a good story about trade-off benefits too, assuming things are not quite as fast.

**Comment by mattpap** [comment]
> PR, are there any performance comparisons between the new options and phantomjs?

I'm not sure if there is anything to add on top of previous discussion. The summary has been that the default will be chrome and firefox will be an alternative if chrome is not available. In terms of startup times, chrome is just slightly slower than phamtomjs and firefox significantly slower (both may improve with the right combination of settings, e.g. using empty profiles). In general speed should be comparable or more likely better, because phantomjs is a very outdated JS runtime.

File: bokeh/io/export.py [hunk 2]
```diff
@@ -159,25 +162,12 @@ def export_svgs(obj, filename=None, height=None, width=None, webdriver=None, tim
 
     return filenames
 
-# this is part of the API for this module
-from .webdriver import terminate_webdriver ; terminate_webdriver
-from .webdriver import webdriver_control ; webdriver_control
```
Comment (bryevdv) [inline bokeh/io/export.py]: @mattpap These breaking changes should have gotten a note in the release migration notes.

File: bokeh/io/export.py [hunk 3]
```diff
@@ -24,35 +24,37 @@
...
 from PIL import Image
 
 # Bokeh imports
+from ..document import Document
 from ..embed import file_html
-from ..resources import INLINE_LEGACY
+from ..models.layouts import LayoutDOM
+from ..resources import INLINE, Resources
 from .util import default_filename
+from .webdriver import WebDriver, webdriver_control
```
Comment (p-himik) [inline bokeh/io/export.py]: @mattpap This import prevents using Bokeh without having geckodriver/chromedriver installed:
```
>>> from bokeh.plotting import figure
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/p-himik/dev/git/bokeh/bokeh/plotting/__init__.py", line 72, in <module>
    from ..io import curdoc; curdoc
  File "/home/p-himik/dev/git/bokeh/bokeh/io/__init__.py", line 23, in <module>
    from .export import export_png, export_svgs
  File "/home/p-himik/dev/git/bokeh/bokeh/io/export.py", line 38, in <module>
    from .webdriver import WebDriver, webdriver_control
  File "/home/p-himik/dev/git/bokeh/bokeh/io/webdriver.py", line 163, in <module>
    webdriver_control = _WebdriverState()
  File "/home/p-himik/dev/git/bokeh/bokeh/io/webdriver.py", line 115, in __init__
    raise RuntimeError("Neither firefox/geckodriver nor chromium-browser/chromedriver are available on system PATH. " \
RuntimeError: Neither firefox/geckodriver nor chromium-browser/chromedriver are available on system PATH. You can install the former with 'conda install -c conda-forge firefox geckodriver'.
```

Comment (mattpap) [reply bokeh/io/export.py]: @p-himik, this should be fixed in PR #9670.

**Pull request merged**